docs/querying/sql-functions.md

```diff
@@ -136,6 +136,42 @@ Returns an array of all values of the specified expression.
 
 Concatenates array inputs into a single array.
 
+## ARRAY_OF_DOUBLES_SKETCH
+
+`ARRAY_OF_DOUBLES_SKETCH(expr, [nominalEntries])`
+
+`ARRAY_OF_DOUBLES_SKETCH(dimensionColumnExpr, metricColumnExpr, ..., [nominalEntries])`
+
+**Function type:** [Aggregation](sql-aggregations.md)
+
+Creates a Tuple sketch.
+
+## ARRAY_OF_DOUBLES_SKETCH_INTERSECT
+
+`ARRAY_OF_DOUBLES_SKETCH_INTERSECT(expr, ..., [nominalEntries])`
+
+**Function type:** [Scalar, sketch](sql-scalar.md#tuple-sketch-functions)
+
+## ARRAY_OF_DOUBLES_SKETCH_NOT
+
+`ARRAY_OF_DOUBLES_SKETCH_NOT(expr, ..., [nominalEntries])`
+
+**Function type:** [Scalar, sketch](sql-scalar.md#tuple-sketch-functions)
+
+## ARRAY_OF_DOUBLES_SKETCH_UNION
+
+`ARRAY_OF_DOUBLES_SKETCH_UNION(expr, ..., [nominalEntries])`
+
+**Function type:** [Scalar, sketch](sql-scalar.md#tuple-sketch-functions)
+
+## ARRAY_OF_DOUBLES_SKETCH_METRICS_SUM_ESTIMATE
```

Comment:
This is quite a long name for a poor user to remember and type.

First, might there be an `ARRAY_OF_LONGS` or `ARRAY_OF_STRINGS` sketch in the future? If so, common SQL practice is to use function overrides: `ARRAY_SKETCH()`, say, which works out the "OF_DOUBLES", etc. part based on the type of the argument. See how `LATEST()` handles this idea.

Second, suppose we'll only ever do an array of doubles. Then, a common approach is to introduce an acronym to shorten names: `AODS`. Or, to make it shorter: "double array sketch" or "DAS".

Then, the suffix only has to be enough to differentiate each function: `DAS_NOT`, `DAS_UNION`, `DAS_SUM`, etc.

Reply:
The set of functions here seem to mimic the new scheme that various Druid principals have outline: functions to create the intermediate types, and more functions to pull final values out of those intermediate types.

That approach makes sense for a native query, where the query spells out the details of aggregation.

For SQL, however, this pattern is not a good fit. SQL provides a single function that goes from inputs to finalized types. For example, for `LATEST()` at the native layer, we may eventually want:

* `LATEST_LONG_SKETCH(string, long) -> COMPLEX<PAIR<string, long>>`
* `LATEST_UNION(COMPLEX<PAIR<string, long>> COMPLEX<PAIR<string, long>>) -> COMPLEX<PAIR<string, long>>`
* `LATEST_LEFT( COMPLEX<PAIR<string, long>>) -> string`
* `LATEST_RIGHT(COMPLEX<PAIR<string, long>> -> long`

However, SQL doesn't provide the semantics to use such functions: one cannot get "inside" of a `GROUP BY` to create the intermediate values, combine them, and then go "outside" of the `GROUP BY` to produce the finalized value.

Instead, in SQL, there is one function, `LATEST(.)` which generates the internal forms when creating the native query.

All of this is a long-winded way of asking, for SQL, do we just need an `APPROX_SUM(ARRAY<DOUBLE>)` function? The rest is worked out by the planner to handle doing approx sums for arrays of doubles?

Reply:
Sketches in the Tuple family are all specialized implementations so I'm not sure making this generic can easily be done, if at all.
Specifically, the various Tuple sketch implementations in the Apache DataSketches project don't seem to share a type hierarchy so determining the type of the serialized sketch at runtime efficiently might be problematic.

Reply:
As to naming— maybe `TUPLE_DOUBLES_SKETCH`, `TUPLE_DOUBLES_INTERSECT`? IMO the `ARRAY_OF` doesn't add much, so we can skip it, and `TUPLE` _does_ add something: the docs call them "tuple sketches" extensively. I'm OK with the originals too. `AODS`, `AODS_INTERSECT` to my eye seem _too_ terse.

As to the bigger potential changes— IMO, the decomposed versions proposed in this patch are a good place to start. Tuple sketches, like theta sketches, are likely to be combined in different ways by different users. Their purpose is to be intersected, not'ed, and unioned in various ways, and then converted to estimates once the set ops are done. Since there isn't a canonical thing that people will largely want to do, it seems like the composable functions are necessary.

Reply:
IMO there are some "pre-composed" functions that would be useful to add, but I don't think we need to do that stuff in this PR, given that we're going to need the decomposed functions anyway.